File: Assets/Bundles/Client/Resources/TS/Controllers/Global/LoadingScreen/LoadingScreenController.lua
-- Compiled with unity-ts v2.1.0-75
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Controller = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Controller
local ClientSignals = require("Client/TS/ClientSignals").ClientSignals
local Game = require("Shared/TS/Game").Game
local Bin = require("Shared/TS/Util/Bin").Bin
local LoadingScreenController
do
	LoadingScreenController = setmetatable({}, {
		__tostring = function()
			return "LoadingScreenController"
		end,
	})
	LoadingScreenController.__index = LoadingScreenController
	function LoadingScreenController.new(...)
		local self = setmetatable({}, LoadingScreenController)
		return self:constructor(...) or self
	end
	function LoadingScreenController:constructor()
		self.coreLoadingScreen = GameObject:Find("CoreLoadingScreen"):GetComponent("CoreLoadingScreen")
		self.coreLoadingScreen:SetProgress("Loading World", 60)
		if Game.LocalPlayer.Character then
			self:FinishLoading()
		else
			self:SetProgress("Waiting for Character", 85)
			local bin = Bin.new()
			bin:Add(ClientSignals.EntitySpawn:Connect(function(event)
				if event.Entity:IsLocalCharacter() then
					bin:Clean()
					self:FinishLoading()
				end
			end))
		end
	end
	function LoadingScreenController:OnStart()
	end
	function LoadingScreenController:SetProgress(step, progress)
		self.coreLoadingScreen:SetProgress(step, progress)
	end
	function LoadingScreenController:FinishLoading()
		self.coreLoadingScreen:Close()
	end
end
-- (Flamework) LoadingScreenController metadata
Reflect.defineMetadata(LoadingScreenController, "identifier", "Bundles/Client/Controllers/Global/LoadingScreen/LoadingScreenController@LoadingScreenController")
Reflect.defineMetadata(LoadingScreenController, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(LoadingScreenController, "$:flamework@Controller", Controller, { {} })
return {
	LoadingScreenController = LoadingScreenController,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Client/Controllers/Global/LoadingScreen/LoadingScreenController.ts
import { Controller, OnStart } from "@easy-games/flamework-core";
import { LoadingScreenController } from "Imports/Core/Client/Controllers/Loading/LoadingScreenController";
import { CoreClientSignals } from "Imports/Core/Client/CoreClientSignals";
import { Game } from "Imports/Core/Shared/Game";
import { Bin } from "Imports/Core/Shared/Util/Bin";
import { WorldAPI } from "Imports/Core/Shared/VoxelWorld/WorldAPI";

@Controller()
export class BWLoadingScreenController implements OnStart {
	constructor(private readonly loadingScreenController: LoadingScreenController) {
		this.CheckWorld();
	}

	private CheckWorld(): void {
		const world = WorldAPI.GetMainWorld();
		if (!world) {
			this.CheckCharacter();
			return;
		}

		if (!world.IsFinishedReplicatingChunksFromServer()) {
			const startTime = os.clock();
			world.OnFinishedReplicatingChunksFromServer.Connect(() => {
				const timeSpent = os.clock() - startTime;
				print("Time spent building world: " + math.floor(timeSpent * 1000) + "ms");
				this.CheckCharacter();
			});
		} else {
			this.CheckCharacter();
		}
	}

	private CheckCharacter(): void {
		if (Game.LocalPlayer.Character) {
			this.loadingScreenController.FinishLoading();
		} else {
			const startTime = os.clock();
			this.loadingScreenController.SetProgress("Waiting for Character", 85);
			const bin = new Bin();
			bin.Add(
				CoreClientSignals.EntitySpawn.Connect((event) => {
					if (event.entity.IsLocalCharacter()) {
						bin.Clean();
						const timeSpent = os.clock() - startTime;
						print("Time spent waiting for character: " + math.floor(timeSpent * 1000) + "ms");
						this.loadingScreenController.FinishLoading();
					}
				}),
			);
		}
	}

	OnStart(): void {}
}
